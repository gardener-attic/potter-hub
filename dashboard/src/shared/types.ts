import * as jsonSchema from "json-schema";
import { IAuthState } from "../reducers/auth";
import { IClusterBomState } from "../reducers/clusterbom";
import { IConfigState } from "../reducers/config";
import { INamespaceState } from "../reducers/namespace";
import { IAppRepositoryState } from "../reducers/repos";
import { IServiceCatalogState } from "../reducers/serviceCatalog";
import { hapi } from "./hapi/release";
import { ResourceKind } from "./ResourceKinds";

// Allow defining multiple error classes
// tslint:disable:max-classes-per-file

class CustomError extends Error {
  // The constructor is defined so we can later on compare the returned object
  // via err.contructor  == FOO
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}
export class ForbiddenError extends CustomError {}
export class UnauthorizedError extends CustomError {}

export class NotFoundError extends CustomError {}

export class ConflictError extends CustomError {}

export class UnprocessableEntity extends CustomError {}

export interface IRepo {
  name: string;
  url: string;
}

export interface IChartVersion {
  id: string;
  attributes: IChartVersionAttributes;
  relationships: {
    chart: {
      data: IChartAttributes;
    };
  };
}

export interface IChartVersionAttributes {
  version: string;
  app_version: string;
  created: string;
}

export interface IChart {
  id: string;
  attributes: IChartAttributes;
  relationships: {
    latestChartVersion: {
      data: IChartVersionAttributes;
    };
  };
}

export interface IChartAttributes {
  name: string;
  description: string;
  home?: string;
  icon?: string;
  keywords: string[];
  maintainers: Array<{
    name: string;
    email?: string;
  }>;
  repo: IRepo;
  sources: string[];
}

export interface IChartState {
  isFetching: boolean;
  selected: {
    error?: Error;
    version?: IChartVersion;
    versions: IChartVersion[];
    readme?: string;
    readmeError?: string;
    values?: string;
    schema?: any;
  };
  deployed: {
    chartVersion?: IChartVersion;
    values?: string;
    schema?: jsonSchema.JSONSchema4;
  };
  items: IChart[];
}

export interface IChartUpdateInfo {
  upToDate: boolean;
  chartLatestVersion: string;
  appLatestVersion: string;
  repository: IRepo;
  error?: Error;
}

export interface IDeployment {
  metadata: {
    name: string;
    namespace: string;
  };
}

export interface IServiceSpec {
  ports: IPort[];
  clusterIP: string;
  type: string;
}

export interface IServiceStatus {
  loadBalancer: {
    ingress?: Array<{ ip?: string; hostname?: string }>;
  };
}

export interface IPort {
  name: string;
  port: number;
  protocol: string;
  targetPort: string;
  nodePort: string;
}

export interface IHTTPIngressPath {
  path: string;
}
export interface IIngressHTTP {
  paths: IHTTPIngressPath[];
}
export interface IIngressRule {
  host: string;
  http: IIngressHTTP;
}

export interface IIngressTLS {
  hosts: string[];
}

export interface IIngressSpec {
  rules: IIngressRule[];
  tls?: IIngressTLS[];
}

export interface IResource {
  apiVersion: string;
  kind: ResourceKind;
  type: string;
  spec: any;
  status: any;
  metadata: {
    name: string;
    namespace: string;
    annotations?: { [key: string]: string };
    creationTimestamp: string;
    deletionTimestamp?: string | null;
    generation?: number;
    labels?: { [key: string]: string };
    resourceVersion: string;
    uid: string;
    selfLink: string;
  };
}

export interface IOwnerReference {
  apiVersion: string;
  blockOwnerDeletion: boolean;
  kind: string;
  name: string;
  uid: string;
}

export interface ISecret {
  apiVersion: string;
  kind: string;
  type: string;
  data: { [s: string]: string };
  metadata: {
    name: string;
    namespace: string;
    annotations: string;
    creationTimestamp: string;
    selfLink: string;
    resourceVersion: string;
    deletionTimestamp?: string;
    uid: string;
  };
}

export interface IDeploymentStatus {
  replicas: number;
  updatedReplicas: number;
  availableReplicas: number;
}

export interface IStatefulsetStatus {
  replicas: number;
  updatedReplicas: number;
  readyReplicas: number;
}

export interface IDaemonsetStatus {
  currentNumberScheduled: number;
  numberReady: number;
}

export interface IRelease extends hapi.release.Release {
  updateInfo?: IChartUpdateInfo;
}

export interface IAppState {
  isFetching: boolean;
  error?: Error;
  deleteError?: Error;
  // currently items are always Helm releases
  items: IRelease[];
  listingAll: boolean;
  listOverview?: IAppOverview[];
  selected?: IRelease;
}

export interface IStoreState {
  serviceCatalog: IServiceCatalogState;
  apps: IAppState;
  auth: IAuthState;
  charts: IChartState;
  clusterBom: IClusterBomState;
  config: IConfigState;
  kube: IKubeState;
  repos: IAppRepositoryState;
  namespace: INamespaceState;
}

interface IK8sResource {
  apiVersion: string;
  kind: string;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#objects */
export interface IK8sObject<M, SP, ST> extends IK8sResource {
  metadata: {
    annotations?: { [key: string]: string };
    creationTimestamp?: string;
    deletionTimestamp?: string | null;
    generation?: number;
    labels?: { [key: string]: string };
    name: string;
    namespace: string;
    resourceVersion?: string;
    uid: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
  spec?: SP;
  status?: ST;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-and-simple-kinds */
export interface IK8sList<I, M> extends IK8sResource {
  items: I[];
  metadata?: {
    resourceVersion?: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
}

export interface IAppRepository
  extends IK8sObject<
    {
      clusterName: string;
      creationTimestamp: string;
      deletionGracePeriodSeconds: string | null;
      deletionTimestamp: string | null;
      resourceVersion: string;
      selfLink: string;
    },
    {
      type: string;
      url: string;
      auth: {
        header: {
          secretKeyRef: {
            name: string;
            key: string;
          };
        };
      };
      resyncRequests: number;
    },
    undefined
  > {}

export interface IAppRepositoryList
  extends IK8sList<
    IAppRepository,
    {
      continue: string;
      resourceVersion: string;
      selfLink: string;
    }
  > {}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#response-status-kind */
export interface IStatus extends IK8sResource {
  kind: "Status";
  status: "Success" | "Failure";
  message: string;
  reason:
    | "BadRequest"
    | "Unauthorized"
    | "Forbidden"
    | "NotFound"
    | "AlreadyExists"
    | "Conflict"
    | "Invalid"
    | "Timeout"
    | "ServerTimeout"
    | "MethodNotAllowed"
    | "InternalError";
  details?: {
    kind?: string;
    name?: string;
    causes?: IStatusCause[] | string;
  };
}

interface IStatusCause {
  field: string;
  message: string;
  reason: string;
}

export interface IRouterPathname {
  router: {
    location: {
      pathname: string;
    };
  };
}

export interface IRuntimeVersion {
  name: string;
  version: string;
  runtimeImage: string;
  initImage: string;
}

export interface IRuntime {
  ID: string;
  versions: IRuntimeVersion[];
  depName: string;
  fileNameSuffix: string;
}

export interface IRBACRole {
  apiGroup: string;
  namespace?: string;
  clusterWide?: boolean;
  resource: string;
  verbs: string[];
}

export interface IAppOverview {
  releaseName: string;
  namespace: string;
  version: string;
  icon?: string;
  status: string;
  chart: string;
  chartMetadata: hapi.chart.Metadata;
  // UpdateInfo is internally populated
  updateInfo?: IChartUpdateInfo;
  description: string;
}

export interface IKubeItem<T> {
  isFetching: boolean;
  item?: T;
  error?: Error;
}

export interface IKubeState {
  items: { [s: string]: IKubeItem<IResource> };
  fetchTargetClusterSecretErr?: Error;
  sockets: { [s: string]: { socket: WebSocket; closeTimer: () => void } };
}

export interface IBasicFormParam {
  path: string;
  type?: jsonSchema.JSONSchema4TypeName | jsonSchema.JSONSchema4TypeName[];
  value?: any;
  title?: string;
  minimum?: number;
  maximum?: number;
  render?: string;
  description?: string;
  enum?: string[];
  hidden?:
    | {
        path: any;
        value: string;
        conditions: Array<{
          path: any;
          value: string;
        }>;
        operator: string;
      }
    | string;
  children?: IBasicFormParam[];
}
export interface IBasicFormSliderParam extends IBasicFormParam {
  sliderMin?: number;
  sliderMax?: number;
  sliderUnit?: string;
}

export interface IApplicationConfig {
  id: string;
  configType: string;
  typeSpecificData: {};
  values: {};
  secretValues: {};
  noReconcile: boolean;
  readyRequirements: IReadyRequirements;
}

export interface IReadyRequirements {
    jobs: IJobReadyRequirement[];
    resources: IResourceReadyRequirement[];
}

export interface IJobReadyRequirement {
    name: string;
    namespace: string
}

export interface IResourceReadyRequirement {
    name: string;
    namespace: string;
    apiVersion: string;
    resource: string;
    fieldPath: string;
    successValues: ISuccessValue[];
}

export interface ISuccessValue {
    value: {};
}

export interface IAutoDelete {
    clusterBomAge: number;
}

export interface IClusterBom
  extends IK8sObject<
    {},
    {
      secretRef: string;
      applicationConfigs: IApplicationConfig[];
      autoDelete: IAutoDelete;
    },
    {
      observedGeneration: number;
      overallState: string;
      overallTime: string;
      conditions: ICondition[];
      applicationStates: IApplicationState[];
      overallNumOfDeployments: number;
      overallNumOfReadyDeployments: number;
      overallProgress: number;
    }
  > {}

export interface ICondition {
  type: string;
  status: string;
  lastUpdateTime: string;
  lastTransitionTime: string;
  reason: string;
  message: string;
}

export interface IApplicationState {
  id: string;
  state: string;
  detailedState: {};
}

export interface IClusterBomList extends IK8sList<IClusterBom, {}> {}

export interface IURLParams {
    targetClusterSecretName: string;
    targetClusterSecretNamespace: string;
}
