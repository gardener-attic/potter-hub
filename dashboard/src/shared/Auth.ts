import Axios, { AxiosResponse } from "axios";
import * as jwt from "jsonwebtoken";
import { K8S_REVERSE_PROXY_URL } from "./Kube";

const AuthTokenKey = "kubeapps_auth_token";
const AuthTokenOIDCKey = "kubeapps_auth_token_oidc";

export const DEFAULT_NAMESPACE = "_all";

export class Auth {
  public static getAuthToken() {
    return localStorage.getItem(AuthTokenKey);
  }

  public static setAuthToken(token: string, oidc: boolean) {
    localStorage.setItem(AuthTokenOIDCKey, oidc.toString());
    return localStorage.setItem(AuthTokenKey, token);
  }

  public static unsetAuthToken() {
    return localStorage.removeItem(AuthTokenKey);
  }

  public static usingOIDCToken() {
    return localStorage.getItem(AuthTokenOIDCKey) === "true";
  }

  public static wsProtocols() {
    const token = this.getAuthToken();
    if (!token) {
      return [];
    }
    return ["base64url.bearer.authorization.k8s.io." + token, "binary.k8s.io"];
  }

  public static fetchOptions(): RequestInit {
    const headers = new Headers();
    headers.append("Authorization", `Bearer ${this.getAuthToken()}`);
    return {
      headers,
    };
  }

  // Throws an error if the token is invalid
  public static async validateToken(token: string) {
    try {
      await Axios.get(K8S_REVERSE_PROXY_URL + "/", {
        headers: { Authorization: `Bearer ${token}` },
      });
    } catch (e) {
      const res = e.response as AxiosResponse;
      if (res.status === 401) {
        throw new Error("invalid token");
      } else if (res.status === 403) {
        throw new Error("token not authorised for cluster");
      }
      // A 403 authorization error only occurs if the token resulted in
      // successful authentication. We don't make any assumptions over RBAC
      // for the root "/" nonResourceURL or other required authz permissions
      // until operations on those resources are attempted (though we may
      // want to revisit this in the future).
      // In the hub project we want to reject everyone with a 403 error
      // since he can't do anything in the ui anyway. We only have
      // admin or unauthorised users.
      if (res.status !== 403) {
        throw new Error(`${res.status}: ${res.data}`);
      }
    }
  }

  // fetchOIDCToken does a HEAD request to collect the Bearer token
  // from the authorization header if exists
  public static async fetchOIDCToken(): Promise<string | null> {
    try {
      const { headers } = await Axios.head("");
      if (headers && headers.authorization) {
        const tokenMatch = (headers.authorization as string).match(/Bearer\s(.*)/);
        if (tokenMatch) {
          return tokenMatch[1];
        }
      }
    } catch (e) {
      // Unable to retrieve token
    }
    return null;
  }

  // defaultNamespaceFromToken decodes a jwt token to return the k8s service
  // account namespace.
  // TODO(mnelson): until we call jwt.verify on the token during validateToken above
  // we use a default namespace for both invalid tokens and tokens without the expected
  // key.
  public static defaultNamespaceFromToken(token: string) {
    const payload = jwt.decode(token);
    const namespaceKey = "kubernetes.io/serviceaccount/namespace";
    if (!payload || !payload[namespaceKey]) {
      return DEFAULT_NAMESPACE;
    }
    return payload[namespaceKey];
  }
}
