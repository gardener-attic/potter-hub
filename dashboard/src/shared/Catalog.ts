export const allRepositories = {
  value: "_all",
  label: "All repositories",
};
