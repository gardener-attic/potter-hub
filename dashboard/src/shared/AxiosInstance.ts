import Axios, { AxiosError, AxiosInstance, AxiosRequestConfig } from "axios";

import { Action, Store } from "redux";
import { ThunkDispatch } from "redux-thunk";
import actions from "../actions";
import { Auth } from "./Auth";
import {
  ConflictError,
  ForbiddenError,
  IStoreState,
  NotFoundError,
  UnauthorizedError,
  UnprocessableEntity,
} from "./types";

export function addAuthHeaders(axiosInstance: AxiosInstance) {
  axiosInstance.interceptors.request.use((config: AxiosRequestConfig) => {
    const authToken = Auth.getAuthToken();
    if (authToken) {
      config.headers.Authorization = `Bearer ${authToken}`;
    }
    return config;
  });
}

export function addErrorHandling(axiosInstance: AxiosInstance, store: Store<IStoreState>) {
  axiosInstance.interceptors.response.use(
    response => response,
    e => {
      const dispatch = store.dispatch as ThunkDispatch<IStoreState, null, Action>;
      const err: AxiosError = e;
      if (
        err.code === undefined &&
        err.message === "Network Error" &&
        !err.response &&
        Auth.usingOIDCToken()
      ) {
        // The OIDC token is no longer valid, logout
        dispatch(actions.auth.expireSession());
      }
      let message = err.message;
      if (err.response && err.response.data.message) {
        message = err.response.data.message;
      }
      switch (err.response && err.response.status) {
        case 401:
          // Global action dispatch to log the user out
          // Expire the session if we are using OIDC tokens
          dispatch(actions.auth.authenticationError(message));
          dispatch(actions.auth.expireSession());
          return Promise.reject(new UnauthorizedError(message));
        case 403:
          return Promise.reject(new ForbiddenError(message));
        case 404:
          return Promise.reject(new NotFoundError(message));
        case 409:
          return Promise.reject(new ConflictError(message));
        case 422:
          return Promise.reject(new UnprocessableEntity(message));
        default:
          return Promise.reject(new Error(message));
      }
    },
  );
}

// Error handling is added with an interceptor in index.tsx
export const axios = Axios.create();
// Authorization headers and error handling are added with an interceptor in index.tsx
const a = Axios.create();
a.defaults.timeout = 1800000;

export const axiosWithAuth = a;
