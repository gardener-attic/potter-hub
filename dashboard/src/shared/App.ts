import { axiosWithAuth } from "./AxiosInstance";
import { hapi } from "./hapi/release";
import { HELM_URL } from "./Kube";
import { IAppOverview, IChartVersion } from "./types";

export class App {
  public static getResourceURL(namespace?: string, name?: string, query?: string) {
    let url = HELM_URL;
    if (namespace) {
      url += `/namespaces/${namespace}`;
    }
    url += "/releases";
    if (name) {
      url += `/${name}`;
    }
    if (query) {
      url += `?${query}`;
    }
    return url;
  }

  public static async create(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = App.getResourceURL(namespace);
    const { data } = await axiosWithAuth.post(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    releaseName: string,
    namespace: string,
    kubeappsNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async rollback(releaseName: string, namespace: string, revision: number) {
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(releaseName: string, namespace: string, keepHistory: boolean) {
    let keepHistoryQuery;
    if (keepHistory) {
      keepHistoryQuery = "keepHistory=true";
    }
    const { data } = await axiosWithAuth.delete(
      App.getResourceURL(namespace, releaseName, keepHistoryQuery),
    );
    return data;
  }

  public static async listApps(namespace?: string, allStatuses?: boolean) {
    let query;
    if (allStatuses) {
      query = "statuses=all";
    }
    const { data } = await axiosWithAuth.get<{ data: IAppOverview[] }>(
      App.getResourceURL(namespace, undefined, query),
    );
    return data.data;
  }

  public static async getRelease(namespace: string, name: string) {
    const { data } = await axiosWithAuth.get<{ data: hapi.release.Release }>(
      this.getResourceURL(namespace, name),
    );
    return data.data;
  }
}
