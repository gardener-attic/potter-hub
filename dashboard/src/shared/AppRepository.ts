import { axiosWithAuth } from "./AxiosInstance";
import { UI_BACKEND_ROOT_URL } from "./Kube";
import { IAppRepository, IAppRepositoryList } from "./types";

export class AppRepository {
  public static async list(namespace: string) {
    const { data } = await axiosWithAuth.get<IAppRepositoryList>(
      AppRepository.getResourceLink(namespace),
    );
    return data;
  }

  public static async get(name: string, namespace: string) {
    const { data } = await axiosWithAuth.get(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async update(name: string, namespace: string, newApp: IAppRepository) {
    const { data } = await axiosWithAuth.put(AppRepository.getSelfLink(name, namespace), newApp);
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const { data } = await axiosWithAuth.delete(AppRepository.getSelfLink(name, namespace));
    return data;
  }

  public static async create(
    name: string,
    namespace: string,
    url: string,
    auth: any,
    syncJobPodTemplate: any,
  ) {
    const { data } = await axiosWithAuth.post<IAppRepository>(
      AppRepository.getResourceLink(namespace),
      {
        apiVersion: "kubeapps.com/v1alpha1",
        kind: "AppRepository",
        metadata: {
          name,
        },
        spec: { auth, type: "helm", url, syncJobPodTemplate },
      },
    );
    return data;
  }

  private static APIBase: string = UI_BACKEND_ROOT_URL;
  private static APIEndpoint: string = `${AppRepository.APIBase}/apprepositories`;
  
  private static getResourceLink(namespace?: string): string {
    return AppRepository.APIEndpoint;
  }

  private static getSelfLink(name: string, namespace: string): string {
    return `${AppRepository.APIEndpoint}/${name}`;
  }
}
