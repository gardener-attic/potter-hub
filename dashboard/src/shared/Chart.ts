import { JSONSchema4 } from "json-schema";
import { axiosWithAuth } from "./AxiosInstance";
import { IChart, IChartVersion } from "./types";
import * as URL from "./url";

export default class Chart {
  public static async fetchCharts(repo: string) {
    const { data } = await axiosWithAuth.get<{ data: IChart[] }>(URL.api.charts.list(repo));
    return data.data;
  }

  public static async fetchChartVersions(id: string) {
    const { data } = await axiosWithAuth.get<{ data: IChartVersion[] }>(
      URL.api.charts.listVersions(id),
    );
    return data.data;
  }

  public static async getChartVersion(id: string, version: string) {
    const { data } = await axiosWithAuth.get<{ data: IChartVersion }>(
      URL.api.charts.getVersion(id, version),
    );
    return data.data;
  }

  public static async getReadme(id: string, version: string) {
    const { data } = await axiosWithAuth.get<string>(URL.api.charts.getReadme(id, version));
    return data;
  }

  public static async getValues(id: string, version: string) {
    const { data } = await axiosWithAuth.get<string>(URL.api.charts.getValues(id, version));
    return data;
  }

  public static async getSchema(id: string, version: string) {
    const { data } = await axiosWithAuth.get<JSONSchema4>(URL.api.charts.getSchema(id, version));
    return data;
  }

  public static async exists(id: string, version: string, repo: string) {
    try {
      await axiosWithAuth.get(URL.api.charts.getVersion(`${repo}/${id}`, version));
    } catch (e) {
      return false;
    }
    return true;
  }

  public static async listWithFilters(name: string, version: string, appVersion: string) {
    const url = `${Chart.APIEndpoint}/charts?name=${name}&version=${encodeURIComponent(
      version,
    )}&appversion=${appVersion}`;
    const { data } = await axiosWithAuth.get<{ data: IChart[] }>(url);
    return data.data;
  }

  private static APIEndpoint: string = "api/chartsvc/v1";
}
