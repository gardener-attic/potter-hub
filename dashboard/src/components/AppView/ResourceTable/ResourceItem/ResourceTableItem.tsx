import * as React from "react";
import { AlertTriangle } from "react-feather";

import { IKubeItem, IResource, ISecret } from "../../../../shared/types";
import LoadingWrapper, { LoaderType } from "../../../LoadingWrapper";
import DaemonSetItemRow from "./DaemonSetItem";
import DeploymentItemRow from "./DeploymentItem";
import SecretItem from "./SecretItem/SecretItem";
import ServiceItem from "./ServiceItem/ServiceItem";
import StatefulSetItemRow from "./StatefulSetItem";

interface IResourceItemProps {
  name: string;
  resource?: IKubeItem<IResource | ISecret>;
  watchResource: () => void;
  closeWatch: () => void;
}

class WorkloadItem extends React.Component<IResourceItemProps> {
  public componentDidMount() {
    this.props.watchResource();
  }

  public componentWillUnmount() {
    this.props.closeWatch();
  }

  public render() {
    const { resource } = this.props;
    return <tr className="flex">{this.renderInfo(resource)}</tr>;
  }

  private renderInfo(resource?: IKubeItem<IResource | ISecret>) {
    const { name } = this.props;
    if (resource === undefined || resource.isFetching) {
      return (
        <React.Fragment>
          <td className="col-3">{name}</td>
          <td className="col-9">
            <LoadingWrapper type={LoaderType.Placeholder} />
          </td>
        </React.Fragment>
      );
    }
    if (resource.error) {
      return (
        <React.Fragment>
          <td className="col-3">{name}</td>
          <td className="col-9">
            <span className="flex">
              <AlertTriangle />
              <span className="flex margin-l-normal">Error: {resource.error.message}</span>
            </span>
          </td>
        </React.Fragment>
      );
    }
    if (resource.item) {
      const r = resource.item as IResource;
      switch (resource.item.kind) {
        case "Deployment":
          return <DeploymentItemRow resource={r} />;
        case "StatefulSet":
          return <StatefulSetItemRow resource={r} />;
        case "DaemonSet":
          return <DaemonSetItemRow resource={r} />;
        case "Service":
          return <ServiceItem resource={r} />;
        case "Secret":
          return <SecretItem resource={resource.item as ISecret} />;
      }
    }
    return null;
  }
}

export default WorkloadItem;
