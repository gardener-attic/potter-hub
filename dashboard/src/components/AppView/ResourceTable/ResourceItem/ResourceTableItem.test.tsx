import { shallow } from "enzyme";
import context from "jest-plugin-context";
import * as React from "react";

import itBehavesLike from "../../../../shared/specs";
import { IKubeItem, IResource } from "../../../../shared/types";
import ResourceTableItem from "./ResourceTableItem";

const kubeItem: IKubeItem<IResource> = {
  isFetching: false,
};

const defaultProps = {
  name: "foo",
  watchResource: jest.fn(),
  closeWatch: jest.fn(),
  statusFields: [],
};

describe("componentDidMount", () => {
  it("calls watchResource", () => {
    const mock = jest.fn();
    shallow(<ResourceTableItem {...defaultProps} watchResource={mock} />);
    expect(mock).toHaveBeenCalled();
  });
});

describe("componentWillUnmount", () => {
  it("calls closeWatch", () => {
    const mock = jest.fn();
    const wrapper = shallow(<ResourceTableItem {...defaultProps} closeWatch={mock} />);
    wrapper.unmount();
    expect(mock).toHaveBeenCalled();
  });
});

context("when fetching resources", () => {
  [undefined, { isFetching: true }].forEach(resource => {
    itBehavesLike("aLoadingComponent", {
      component: ResourceTableItem,
      props: {
        ...defaultProps,
        resource,
      },
    });

    it("displays the name of the resource", () => {
      const wrapper = shallow(<ResourceTableItem {...defaultProps} resource={resource} />);
      expect(wrapper.text()).toContain("foo");
    });
  });
});

context("when there is an error fetching the resource", () => {
  const resource = {
    error: new Error('deployments "foo" not found'),
    isFetching: false,
  };
  const wrapper = shallow(<ResourceTableItem {...defaultProps} resource={resource} />);

  it("diplays the resource name in the first column", () => {
    expect(
      wrapper
        .find("td")
        .first()
        .text(),
    ).toEqual("foo");
  });

  it("displays the error message in the second column", () => {
    expect(
      wrapper
        .find("td")
        .at(1)
        .text(),
    ).toContain('Error: deployments "foo" not found');
  });
});

context("when there is a valid resouce", () => {
  it("renders info about the resource status", () => {
    const deployment = {
      metadata: {
        name: "foo",
      },
      status: { replicas: 1, updatedReplicas: 1, availableReplicas: 1 },
    } as IResource;
    kubeItem.item = deployment;
    const wrapper = shallow(
      <ResourceTableItem {...defaultProps} resource={kubeItem} name={deployment.metadata.name} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
