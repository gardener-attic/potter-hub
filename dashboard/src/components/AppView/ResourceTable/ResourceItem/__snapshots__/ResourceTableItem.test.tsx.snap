// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when fetching resources loading spinner matches the snapshot 1`] = `
<tr
  className="flex"
>
  <td
    className="col-3"
  >
    foo
  </td>
  <td
    className="col-9"
  >
    <LoadingWrapper
      loaded={false}
      type={1}
    />
  </td>
</tr>
`;

exports[`when fetching resources loading spinner matches the snapshot 2`] = `
<tr
  className="flex"
>
  <td
    className="col-3"
  >
    foo
  </td>
  <td
    className="col-9"
  >
    <LoadingWrapper
      loaded={false}
      type={1}
    />
  </td>
</tr>
`;

exports[`when there is a valid resouce renders info about the resource status 1`] = `
<tr
  className="flex"
/>
`;
